File: quickSort/quick_sort.rb
def quick_sort(arraylist)
  return if arraylist.nil? || arraylist.empty?
  quick_helper(arraylist, 0, arraylist.length - 1)
end

def quick_helper(arraylist, first, last)
  return if first >= last
  left = first; right = last
  pivot = (first + last) / 2
  while (left < right)
    while left <= right && arraylist[left] <= arraylist[pivot]
      left += 1
    end
    while left <= right && arraylist[right] >= arraylist[pivot]
      right -= 1
    end
    if left <= right
      arraylist[left], arraylist[right] = arraylist[right], arraylist[left]
      left += 1; right -= 1
    end
  end
  quick_helper(arraylist, first, right)
  quick_helper(arraylist, left, last)
end

puts arr = [5, 4, 3, 2, 1]
quick_sort(arr)
puts arr

File: quickSort/quickSort.java
// package whatever; // don't place package name!

import java.io.*;
import java.util.*;

class solution {
  public static void quickSort(int[] arr) {
    if (arr == null || arr.length == 0) {
      return;
    }
    quickHelper(arr, 0, arr.length - 1);
  }

  public static void quickHelper(int[] arr, int start, int end) {
    if (start > end) {
      return;
    }
    int left = start;
    int right = end;
    int pivot = (start + end) / 2;
    while (left <= right) {
      while (left <= right && arr[left] < arr[pivot]) {
        left++;
      }
      while (left <= right && arr[right] > arr[pivot]) {
        right--;
      }
      if (left <= right) {
        int temp = arr[right];
        arr[right] = arr[left];
        arr[left] = temp;
        left++;
        right--;
      }
    }
    quickHelper(arr, left, end);
    quickHelper(arr, start, right);
  }
  public static void main(String[] args) {
    int[] arr = new int[]{5, 4, 3, 2, 1};
    quickSort(arr);
    System.out.println(Arrays.toString(arr));
  }
}
